drivers/console/semihosting_console.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020 Maciej Sopyło <maciek134@gmail.com>
+ * Copyright (c) 2018 Josef Gajdusek <atx@atx.name>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <sys/printk.h>
+#include <device.h>
+#include <init.h>
+#include <string.h>
+
+extern void __printk_hook_install(int (*fn)(int));
+extern void __stdout_hook_install(int (*fn)(int));
+
+static char buffer[CONFIG_SEMIHOSTING_BUFFER_SIZE];
+static u32_t buffer_pos;
+
+static void semihosting_flush_buffer(struct k_timer *timer);
+static K_TIMER_DEFINE(send_timer, semihosting_flush_buffer, NULL);
+
+
+static void semihosting_flush_buffer(struct k_timer *timer)
+{
+	if (buffer[buffer_pos] != 0) {
+		buffer[buffer_pos++] = 0;
+	}
+
+	register int r0 __asm__("r0") = 0x04; // SYS_WRITE0
+	register char *r1 __asm__("r1") = buffer;
+	__asm__ volatile("bkpt $0xab"
+			:
+			: "r" (r0), "r" (r1));
+
+	buffer_pos = 0;
+}
+
+
+static int semihosting_console_out(int character)
+{
+	int key = irq_lock();
+	bool end;
+
+	buffer[buffer_pos++] = character;
```

Comment:
Should validate `buffer_pos` before referencing.

Reply:
If I understand correctly how `k_timer_status_sync` works it shouldn't be necessary - `0 <= buffer_pos < ARRAY_SIZE(buffer) - 1` should always be true?

Reply:
Fair enough, this is not safety critical code after all.